File: src/cc_proxy/claude_cli.py
"""Claude CLI subprocess management and stream-json parsing."""

import asyncio
import json
import time
from collections.abc import AsyncIterator

from .models import SendMessageResult, StreamEvent

_session_locks: dict[str, asyncio.Lock] = {}


def get_session_lock(session_id: str) -> asyncio.Lock:
    if session_id not in _session_locks:
        _session_locks[session_id] = asyncio.Lock()
    return _session_locks[session_id]


async def send_message(
    message: str,
    session_id: str,
    cwd: str,
    claude_cli_path: str = "claude",
    timeout: int = 300,
) -> AsyncIterator[StreamEvent]:
    """Send a message to a Claude Code session via CLI subprocess.

    Yields StreamEvent objects as they arrive from the CLI stream-json output.
    The final event will have is_final=True.
    """
    cmd = [
        claude_cli_path,
        "-p", message,
        "--resume", session_id,
        "--output-format", "stream-json",
        "--verbose",
        "--dangerously-skip-permissions",
    ]

    start = time.monotonic()

    proc = await asyncio.create_subprocess_exec(
        *cmd,
        cwd=cwd,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )

    try:
        async for line in _read_lines(proc.stdout):
            event = _parse_stream_line(line)
            if event is not None:
                yield event
                if event.is_final:
                    return

        await asyncio.wait_for(proc.wait(), timeout=10)
    except asyncio.TimeoutError:
        proc.kill()
        yield StreamEvent(event_type="error", content="CLI timeout", is_final=True)
    finally:
        elapsed = int((time.monotonic() - start) * 1000)
        if proc.returncode is None:
            proc.kill()


async def create_new_session(
    cwd: str,
    name: str | None = None,
    claude_cli_path: str = "claude",
) -> tuple[str, str]:
    """Create a new Claude Code session. Returns (session_id, response_text)."""

    cmd = [
        claude_cli_path,
        "-p", "Session initialized.",
        "--output-format", "stream-json",
        "--dangerously-skip-permissions",
    ]
    if name:
        cmd.extend(["--name", name])

    proc = await asyncio.create_subprocess_exec(
        *cmd,
        cwd=cwd,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )

    result_text = ""
    result_session_id = ""
    async for line in _read_lines(proc.stdout):
        try:
            data = json.loads(line.strip())
        except json.JSONDecodeError:
            continue
        if data.get("type") == "result":
            result_text = data.get("result", "")
            result_session_id = data.get("session_id", "")

    await proc.wait()
    return result_session_id, result_text


async def _read_lines(stream: asyncio.StreamReader | None) -> AsyncIterator[str]:
    if stream is None:
        return
    while True:
        line = await stream.readline()
        if not line:
            break
        yield line.decode("utf-8", errors="replace")


def _parse_stream_line(line: str) -> StreamEvent | None:
    try:
        data = json.loads(line.strip())
    except json.JSONDecodeError:
        return None

    msg_type = data.get("type")

    if msg_type == "result":
        return StreamEvent(
            event_type="result",
            content=data.get("result", ""),
            is_final=True,
        )

    if msg_type == "assistant":
        content_blocks = data.get("message", {}).get("content", [])
        texts = [
            b.get("text", "") for b in content_blocks
            if isinstance(b, dict) and b.get("type") == "text"
        ]
        if texts:
            return StreamEvent(
                event_type="assistant_text",
                content="".join(texts),
            )

    return None

File: src/claude-cli.ts
import { spawn, type ChildProcess } from "node:child_process";
import type { StreamEvent } from "./types.js";
import { resolveAuthEnv } from "./env-resolver.js";

const sessionLocks = new Map<string, Promise<void>>();
const STDIO: ["ignore", "pipe", "pipe"] = ["ignore", "pipe", "pipe"];

function getSessionLock(sessionId: string): { current: Promise<void>; resolve: () => void } {
  let resolve!: () => void;
  const current = sessionLocks.get(sessionId) ?? Promise.resolve();
  const next = new Promise<void>((r) => {
    resolve = r;
    setTimeout(() => { try { r(); } catch {} }, 60_000);
  });
  const chained = current.then(() => next);
  sessionLocks.set(sessionId, chained);
  return { current: chained, resolve };
}

function buildSpawnOpts(cwd: string): Parameters<typeof spawn>[2] {
  return {
    cwd,
    env: { ...process.env, ...resolveAuthEnv() },
    stdio: STDIO,
  };
}

interface SendMessageOptions {
  message: string;
  sessionId: string;
  cwd: string;
  claudeCliPath?: string;
  timeout?: number;
  planMode?: boolean;
  forkSession?: boolean;
}

export function sendMessage({
  message,
  sessionId,
  cwd,
  claudeCliPath = "claude",
  timeout = 300,
  planMode = false,
  forkSession = false,
}: SendMessageOptions): Promise<StreamEvent[]> {
  const { current: lock, resolve } = getSessionLock(sessionId);

  return lock.then(() => {
    const args = [
      "-p", message,
      "--resume", sessionId,
      "--output-format", "stream-json",
      "--verbose",
      "--dangerously-skip-permissions",
    ];
    if (planMode) args.push("--permission-mode", "plan");
    if (forkSession) args.push("--fork-session");

    const proc = spawn(claudeCliPath, args, buildSpawnOpts(cwd));

    return collectStream(proc, timeout).finally(() => {
      resolve();
      killProc(proc);
    });
  });
}

interface CreateSessionOptions {
  cwd: string;
  name?: string;
  claudeCliPath?: string;
  timeout?: number;
}

export async function createNewSession({
  cwd,
  name,
  claudeCliPath = "claude",
  timeout = 300,
}: CreateSessionOptions): Promise<{ sessionId: string; response: string }> {
  const args = [
    "-p", "Session initialized.",
    "--output-format", "stream-json",
    "--verbose",
    "--dangerously-skip-permissions",
  ];
  if (name) args.push("--name", name);

  const proc = spawn(claudeCliPath, args, buildSpawnOpts(cwd));
  const events = await collectStream(proc, timeout);

  const resultEvent = events.find((e) => e.event_type === "result");
  return {
    sessionId: resultEvent?.session_id ?? "",
    response: resultEvent?.content ?? "",
  };
}

function collectStream(proc: ChildProcess, timeoutSec: number): Promise<StreamEvent[]> {
  return new Promise((resolve) => {
    const events: StreamEvent[] = [];
    let buffer = "";
    let stderrText = "";
    let resolved = false;

    const timer = setTimeout(() => {
      if (!resolved) {
        resolved = true;
        events.push({ event_type: "error", content: "CLI timeout", is_final: true });
        killProc(proc);
        resolve(events);
      }
    }, timeoutSec * 1000);

    proc.stderr?.on("data", (d) => { stderrText += d.toString(); });

    proc.stdout?.on("data", (d) => {
      if (resolved) return;
      buffer += d.toString();
      const lines = buffer.split("\n");
      buffer = lines.pop() ?? "";

      for (const line of lines) {
        const event = parseStreamLine(line);
        if (event) {
          events.push(event);
          if (event.is_final) {
            resolved = true;
            clearTimeout(timer);
            killProc(proc);
            resolve(events);
            return;
          }
        }
      }
    });

    proc.on("close", (code) => {
      if (resolved) return;
      resolved = true;
      clearTimeout(timer);

      if (buffer.trim()) {
        const event = parseStreamLine(buffer);
        if (event) events.push(event);
      }

      if (code !== 0 && code !== null && events.length === 0) {
        events.push({ event_type: "error", content: stderrText || `CLI exited with code ${code}`, is_final: true });
      }

      resolve(events);
    });
  });
}

function parseStreamLine(line: string): StreamEvent | null {
  let data: Record<string, unknown>;
  try {
    data = JSON.parse(line.trim());
  } catch {
    return null;
  }

  if (data.type === "result") {
    const sessionId = (data.session_id as string) ?? undefined;
    if (data.is_error) {
      const errors = data.errors as string[] | undefined;
      return {
        event_type: "error",
        content: errors?.join("; ") || (data.result as string) || "CLI error",
        is_final: true,
        session_id: sessionId,
      };
    }
    return {
      event_type: "result",
      content: (data.result as string) ?? "",
      is_final: true,
      session_id: sessionId,
    };
  }

  if (data.type === "assistant") {
    const msg = data.message as Record<string, unknown> | undefined;
    const content = (msg?.content ?? []) as Record<string, unknown>[];
    const texts = content
      .filter((b) => b.type === "text")
      .map((b) => (b.text as string) ?? "");
    if (texts.length > 0) {
      return { event_type: "assistant_text", content: texts.join(""), is_final: false };
    }
  }

  return null;
}

function killProc(proc: ChildProcess) {
  if (proc.exitCode === null) proc.kill();
}
